### PR TITLE
Add support for autocomplete based on a prefix

### LINKED ISSUE #12: textDocument/complete: Use "isIncomplete" property
Based on some anecdotal evidence and common sense the server should not be sending more than ~100 suggestions for completion in one response - instead it should just send the first 100 best matches and set `isIncomplete` to `true` and let the client recompute (effectively force the user to keep typing to filter suggestions).

The LSP doesn't define a workflow for passing the letters for further filtering of the candidate list, so we will need to read it from the config, basically read anything on the left side of the position until "enclosing character". It may not be ideal in terms of performance, but probably better for simplicity to just leverage HCL parser there to understand where does the content end.

### PR DESCRIPTION
**Pull request by njuCZ**
Related: #12 
- in your test case, when `Line:   2, Column: 5`, the Byte should be `21` but not `20`
- add some corner case for func `TokenAtPosition`
- I just filter the results in the last step, please confirm whether we should filter the candidates before putting them into the candidate list
- tested in vscode, it's ok for this feature

### CODE REVIEW

**Comment by radeksimko** [comment]
> in your test case, when Line: 2, Column: 5, the Byte should be 21 but not 20

Agreed - I missed that one, thank you for spotting and fixing it 👍 

> add some corner case for func TokenAtPosition

yep - I can work on that and push it to your branch, or in a separate PR

> I just filter the results in the last step, please confirm whether we should filter the candidates before putting them into the candidate list

That's a good question, I left a comment inline about that - I think we should indeed do it before.

> tested in vscode, it's ok for this feature

Me and @paultyng also tested it in VSCode ourselves and it works well except for one edge case, which is completion in the middle of an incomplete label, e.g.

```hcl
resource "aws_<HERE>" "name" {
  
}
```

I will look into this ^

File: langserver/handlers/complete.go [hunk 1]
```diff
@@ -54,10 +54,16 @@ func (h *logHandler) TextDocumentComplete(ctx context.Context, params lsp.Comple
 	p.SetLogger(h.logger)
 	p.SetSchemaReader(sr)
 
+	token, hclPos, err := hclFile.TokenAtPosition(fPos)
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply langserver/handlers/complete.go:57]: Relatedly we may need to change the interface of `configBlockFactory.New()` to take tokens instead of `hclsyntax.Block` and remove `parser.ParseBlockFromHCL` in favour of `parser.ParseBlockFromTokens(hcl.Tokens)`, so we retain the tokens in the parser, which in turn leads us to a conclusion that the original abstraction of `TokenAtPosition` doesn't work very well here and should probably be reimplemented as a replacement of `BlockAtPosition(fPos) hcl.Block`, e.g. `BlockTokensAtPosition() hclsyntax.Tokens`.

Then finally we can re-implement the original abstraction inside `internal/terraform/lang` as `TokenAtPos(hclsyntax.Tokens, hcl.Pos) hclsyntax.Token`.

[the remark this answers is not in the document]
Comment (radeksimko) [reply langserver/handlers/complete.go:57]: I will attempt to do the refactoring outside of this PR and see how far I can get.

[the remark this answers is not in the document]
Comment (radeksimko) [reply langserver/handlers/complete.go:57]: Actually bytes may not be the best format for passing around, since we loose context like position within the whole file. 🤔 I'll try to figure out something with tokens instead.

**Comment by radeksimko** [comment]
That previously discussed refactoring is now ready for review in https://github.com/hashicorp/terraform-ls/pull/125

**Comment by radeksimko** [comment]
#125 was merged, so this PR now may need rebasing. Let me know if you need any further help here.

**Comment by njuCZ** [comment]
@radeksimko to filter candidates before putting  into the candidate list, I have done a little refactor of the codes, Could you please have a look ?

**Comment by paultyng** [comment]
Should we remove this check now? As it stands, I can't "recomplete" an attribute if I chose the wrong name from the middle, just the begining:

https://github.com/hashicorp/terraform-ls/pull/119/files#diff-7f004988141686ad2e40b48a8346a3b7R76-R79

**Comment by radeksimko** [comment]
I've done some further refactoring as this functionality proved that the original interface/abstraction with pure tokens wasn't really sufficient.

The main motivation behind the refactoring here was to reduce and simplify the top layer (handler), decouple some generic HCL parsing logic and finally to keep `CompletionCandidate` immutable (as opposed to letting anyone `SetPrefix`) and allow passing around already "trimmed" candidates as opposed to letting the `lsp` package handle the trimming via `PrefixRange`.

I also put back some tests which were removed, as I think there's still value in keeping them and testing the new behaviour (error).
